toonz/sources/toonz/commandbar.cpp

```diff
@@ -1,84 +1,196 @@
-
+﻿
 
 #include "commandbar.h"
 
 // Tnz6 includes
 #include "tapp.h"
 #include "menubarcommandids.h"
+#include "tsystem.h"
+#include "commandbarpopup.h"
+
 // TnzQt includes
+#include "toonzqt/menubarcommand.h"
 #include "toonzqt/gutil.h"
 
 // TnzLib includes
-#include "toonz/preferences.h"
 #include "toonz/tscenehandle.h"
-#include "toonzqt/menubarcommand.h"
-
+#include "toonz/toonzscene.h"
+#include "toonz/childstack.h"
+#include "toonz/toonzfolders.h"
 // Qt includes
 #include <QWidgetAction>
-
+#include <QXmlStreamReader>
+#include <QtDebug>
+#include <QMenuBar>
+#include <QContextMenuEvent>
 
 //=============================================================================
 // Toolbar
 //-----------------------------------------------------------------------------
 
 #if QT_VERSION >= 0x050500
 CommandBar::CommandBar(QWidget *parent, Qt::WindowFlags flags,
-                             bool isCollapsible)
+                       bool isCollapsible, bool isXsheetToolbar)
 #else
-CommandBar::CommandBar(XsheetViewer *parent, Qt::WFlags flags)
+CommandBar::CommandBar(QWidget *parent, Qt::WFlags flags)
 #endif
     : QToolBar(parent), m_isCollapsible(isCollapsible) {
   setObjectName("cornerWidget");
   setObjectName("CommandBar");
+  fillToolbar(this, isXsheetToolbar);
+}
 
-  TApp *app        = TApp::instance();
-  m_keyFrameButton = new ViewerKeyframeNavigator(this, app->getCurrentFrame());
-  m_keyFrameButton->setObjectHandle(app->getCurrentObject());
-  m_keyFrameButton->setXsheetHandle(app->getCurrentXsheet());
+//-----------------------------------------------------------------------------
 
-  QWidgetAction *keyFrameAction = new QWidgetAction(this);
-  keyFrameAction->setDefaultWidget(m_keyFrameButton);
+void CommandBar::updateEditInPlaceStatus() {
+  TApp *app         = TApp::instance();
+  ToonzScene *scene = app->getCurrentScene()->getScene();
+  int ancestorCount = scene->getChildStack()->getAncestorCount();
+  if (ancestorCount == 0) {
+    m_editInPlace->setChecked(false);
+    return;
+  }
+  m_editInPlace->setChecked(scene->getChildStack()->getEditInPlace());
+}
```

Comment:
In many part of this PR you are trying to synchronize the state of `Edit In Place` property.

I'm so sorry for not coming to realize when I merged the `Toggle Edit in Place` command PR ( #306 )...,
but now I wonder why this toggle is stored in the scene data despite the fact that (as far as I can see) it seems to affect only display behavior of the viewer.

If this toggle is related only to the interface, then how about bringing this value out from `ChildStack` and making it one of the viewer toggle (just like `MI_ViewTable`, `MI_ICheck`, etc. ) by using `MainWindow::createToggle()` instead of `MainWindow::createMenuXsheetAction()` ? 
I think it will make it easier to handle this command in the command bar.

Anyway, making this command easily accessible via command bar will be helpful especially for cut-out animators. Thanks!